#30 output 처리 및 input parsing 완료

Naing0126 opened the pull request.

Naing0126 committed:
fix ClustererFileInput undefine error in file_io.cc

Naing0126 committed:
Implement output metadata file from tracklets

Naing0126 committed:
Implement mapping file

Naing0126 committed:
Change get_output_path() method's name to output_path()

Naing0126 committed:
Implement clusterer_option

Naing0126 committed:
Implement parsing input data

Naing0126 committed:
remove temp file

Naing0126 committed:
Implement parsing mapping file

kyukyukyu commented on src/clusterer/main.cc:
```diff
@@ -7,7 +7,7 @@ int main(int argc, const char** argv) {
     ugproj::Configuration cfg;
     if (cfg.load(argc, argv)) {
-      throw new std::exception();
+      return 1;
```
예외 대신 `return 1;`을 바로 때린 이유가 무엇인가요?

Naing0126 replied:
아무 이유 없슴다.. 다시 그래도 예외로 바꾸겠습니다

kyukyukyu commented on src/clusterer/main.cc:
```diff
@@ -58,6 +59,7 @@ int main(int argc, const char** argv) {
     ugproj::FaceClustersVisualizer visualizer(&writer);
     visualizer.visualize(tracked_positions, tracklets,
                          cfg.clustering.k, cluster_ids);
+     */
```
이 부분 주석처리 한 이유가 무엇인가요?

kyukyukyu commented on src/file_io.cc:
```diff
@@ -22,6 +22,86 @@ cv::CascadeClassifier& TrackerFileInput::cascade() {
   return this->cascade_;
 }
 
+int ClustererFileInput::open(const Configuration& cfg) {
```
Global metadata 읽는 루틴과 tracklet (tracklet metadata 포함해서) 읽는 루틴을 `private` 메서드로 정의한 다음에 이 메서드 안에서 호출하는 식으로 코드를 짜면 더 나을 것 같습니다. 메서드 하나가 너무 길어요.

kyukyukyu commented on src/file_io.cc:
```diff
@@ -22,6 +22,86 @@ cv::CascadeClassifier& TrackerFileInput::cascade() {
+  // Parsing metadata file and mapping file to tracklet and tracked_positions
+
+  // mapping file parsing to tracked_positions_
+  char mappingFilename[256];
```
여러 단어로 구성된 변수 이름에는 camelCase 대신 underscore를 사용해야 합니다. e.g. `mapping_filename`

kyukyukyu commented on src/file_io.cc:
```diff
@@ -22,6 +22,86 @@ cv::CascadeClassifier& TrackerFileInput::cascade() {
+
+  // mapping file parsing to tracked_positions_
+  char mappingFilename[256];
+  std::sprintf(mappingFilename,"%s",cfg.mapping_filepath.string().c_str());
```
코드 쭉 보니까 여기 이후로 `mappingFilename`을 수정하는 코드가 없는 거 같은데... `std::sprintf()`를 호출해 버리면 불필요한 연산을 하게 되죠.
그리고 `boost::filesystem::path` 객체를 문자열로 변환할 때에는 `string()` 메서드 대신 `native()`를 쓰는 것이 안전합니다.
정리하면, 아래와 같이 하면 더 나을 것 같아요.

``` cpp
const auto* mappingFilename = cfg.mapping_filepath.native().c_str();
```

kyukyukyu commented on src/file_io.cc:
```diff
@@ -22,6 +22,86 @@ cv::CascadeClassifier& TrackerFileInput::cascade() {
+      
+      this->tracklets_.push_back(ugproj::FaceTracklet(i-1));
+      
+      std::string trackletFilename = cfg.input_dirpath.string();
```
String concatenation 대신에 `std::sprintf()`를 사용하는 것을 권합니다.

kyukyukyu commented on src/tracker/face_tracker.cc:
```diff
@@ -466,31 +477,64 @@ int FaceTracker::write_tracklet(
   return this->writer_->write_image(img_tracklet, filename);
 }
 
+int FaceTracker::write_mapping_file(
```
`ugproj::FileWriter` 클래스에 YAML 파일을 쓰는 메서드를 정의하여 `ugproj::FaceTracker::write_mapping_file()` 및 `ugproj::FaceTracker::write_tracklet_metadata()`에서 사용하는 형태로 구현하는 것이 바람직합니다. (e.g. `ugproj::FileWriter::write_yaml(filename, data)`) 그렇게 하면 file output 자체를 `ugproj::FileWriter` 클래스 객체가 처리할 수 있습니다. 갑자기 `output_path()` 같이 불필요한 getter method를 만들 필요도 없고요.

`ugproj::FileWriter::write_image()`와 마찬가지로, 이 메서드는 파일의 이름과 파일에 쓸 데이터를 매개변수로 받아야 합니다. [여기](http://docs.opencv.org/modules/core/doc/xml_yaml_persistence.html#filestorage-root)에서 알 수 있듯이, 이 데이터는 mapping 형태이며, OpenCV에서는 이를 위해 `cv::FileNode` 타입을 제공합니다. 그래서, `ugproj::FileWriter::write_yaml(filename, data)`에서 `data`의 타입은 `const cv::FileNode&`로 정하는 것이 좋습니다.

kyukyukyu commented on src/structure.cc:
```diff
@@ -39,9 +39,10 @@ int Configuration::load(int argc, const char** argv) {
   namespace po = boost::program_options;
 
   try {
-    po::options_description generic_options("Generic options");
```
[여기](https://github.com/kyukyukyu/ugproj-detector/issues/31#issuecomment-141640630)에서 command line option은 두 프로그램이 똑같이 갖도록 정했습니다.

kyukyukyu commented on src/structure.cc:
```diff
@@ -59,6 +60,34 @@ int Configuration::load(int argc, const char** argv) {
          ->default_value("output"),
        "path to output directory.")
     ;
+    clusterer_options.add_options()
```
Clusterer를 위한 command line options을 따로 정의할 필요가 없는 이유는 이렇습니다.
- Tracklet metadata filename의 형식이 이슈 #30 에서 정해졌습니다.
- (정해진 적은 없지만) Global metadata filename 또한 파일이름을 정해두면 옵션으로 받을 필요가 없어집니다.
- Clusterer의 input directory가 곧 tracker의 output directory입니다. 그 말인즉슨, `--output-dir` 옵션으로 주어진 디렉토리 경로를 그대로 사용하면 됩니다.
- Tracklet metadata filename의 형식과 global metadata filename이 정해져 있고, clusterer의 input directory가 주어지기 때문에 `--metadata-file`과 `--mapping-file` 옵션의 값을 사용자로부터 입력받을 필요가 없습니다.

kyukyukyu commented on src/structure.cc:
```diff
@@ -172,8 +201,11 @@ int Configuration::load(int argc, const char** argv) {
     po::options_description visible_options("Allowed options");
-    visible_options.add(generic_options);
-
+    if(!std::string("./tracker").compare(argv[0])){
```
...어쩔 수 없었다는 건 이해하지만, 이런 코드는 작성하면 안 됩니다. 만약에 실행 파일 이름이 바뀌기라도 한다면? 프로그램을 실행 파일이 있는 경로가 아닌 다른 경로에서 실행한다면? 차라리 전처리기 매크로를 사용하세요.

kyukyukyu commented on src/structure.h:
```diff
@@ -137,6 +140,9 @@ class Face {
     Face(const temp_idx_t frameIndex, const cv::Rect& rect,
                   const cv::Mat& image) :
         frameIndex(frameIndex), rect(rect), image(image), tracklet_id(0) {};
+    Face(const temp_idx_t frameIndex, const cv::Mat& image,
```
Good job

kyukyukyu commented on src/tracker/face_tracker.cc:
```diff
@@ -97,6 +97,11 @@ int FaceTracker::track(std::vector<unsigned long>* tracked_positions,
       break;
     }
 
+    if(pos > 1500){
```
이이거어 하아지이 마아세에요오...

kyukyukyu commented on src/file_io.cc:
```diff
@@ -22,6 +22,86 @@ cv::CascadeClassifier& TrackerFileInput::cascade() {
+      cv::Mat trackletMat = cv::imread(trackletFilename,CV_LOAD_IMAGE_COLOR);
+      
+      if(!trackletMat.data){
+          std::printf("Could not open or find the image\n");
```
오류 메시지는 `stderr`로 출력해야 합니다. 그리고 formatting이 필요없는 문자열은 `std::fputs()`로 출력하는 것이 좋습니다.

``` cpp
std::fputs("Could not open or find the image\n", stderr);
```

kyukyukyu commented on src/file_io.cc:
```diff
@@ -22,6 +22,86 @@ cv::CascadeClassifier& TrackerFileInput::cascade() {
+
+      (*it2)["frame_indices"] >> frame_indices;
+
+      static const int kSize = 64;
```
이슈 #30 의 '설정 파일에 옵션 추가' 섹션을 참조하세요. 설정 파일에서 입력 받기로 했습니다.

Naing0126 committed:
Implement dimensionality reduction with pca

kyukyukyu committed:
Fix whitespaces

kyukyukyu committed:
Fix bugs in main() for clusterer

kyukyukyu committed:
Fix bugs in FaceClusterer

* std::vector::reserve() does not create elements. To populate elements,
  constructor should be used, or just std::vector::push_back().

kyukyukyu committed:
Fix bugs in visualizer for clustering result

* As described in last commit, std::vector::reserve() does not create
  any element. Instead of this, constructor for std::vector is used.
* Face objects created in clusterer does not have rect information.
  Calling Face::resized_image() causes crash since scale factor becomes
  +inf.

kyukyukyu committed:
Fix a bug in ClustererFileInput

* Invalid index number.

kyukyukyu committed:
Remove code for jumping frames

* Please, 제발, DO NOT PUT CODE LIKE THIS.

kyukyukyu committed:
Increase the reduced dimensionality from 2 to 64

kyukyukyu committed:
Update sample configuration file

kyukyukyu committed:
Resolve #38

kyukyukyu committed:
Add missing OpenCV dependency to README.md

kyukyukyu committed:
Resolve #39

kyukyukyu committed:
Remove duplicated code for drawing multiple faces

kyukyukyu committed:
Use image's size in ugproj::Face::resized_image()

* Using the size of rect causes an error if the rect is not initialized.

kyukyukyu committed:
Add configuration section for output

kyukyukyu committed:
Remove unused parameter in writing tracklet
